File: java/DivideAndConquer/QuickSort.java
public class QuickSort{
    public static void main(String args[]){
        int arr[] = {6, 3, 9, 8, 2, 5};
        quickSort(arr, 0, arr.length-1);
        printArr(arr);
    }

    public static void printArr(int arr[]){
        for(int i=0; i<arr.length; i++){
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }

    public static void quickSort(int arr[], int si, int ei){
        if(si >= ei){
            return;
        }

        // select a pivot (we take left here)
        
        // partitioning recursion
        int partitionIdx = partition(arr, si, ei);
        quickSort(arr, si, partitionIdx-1);     // left array
        quickSort(arr, partitionIdx+1, ei);     // right array
    }

    public static int partition(int arr[], int si, int ei){

        // we select the pivot as last element 
        int pivot = arr[ei];
        int i = si-1;           // makes space for elements smaller than pivot
        
        for(int j=si; j<ei; j++){
            if(pivot >= arr[j]){
                i++;
                // swap i and j
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
                
            }
        }

        // incremented i again for maiking space for pivot
        i++;
        int temp = pivot;
        arr[ei] = arr[i];
        arr[i] = temp;
        return i;
    }
}

File: c++/Array/QuickSort.cpp
#include <iostream>
#include <vector>
using namespace std;

// Function to partition the array
int partition(vector<int> &arr, int low, int high)
{
    int pivot = arr[high]; // choose the last element as pivot
    int i = low - 1;       // pointer for smaller element

    for (int j = low; j < high; j++)
    {
        // if current element is smaller than or equal to pivot
        if (arr[j] <= pivot)
        {
            i++;
            swap(arr[i], arr[j]);
        }
    }

    // place pivot in correct position
    swap(arr[i + 1], arr[high]);
    return i + 1;
}

// Recursive QuickSort function
void quickSort(vector<int> &arr, int low, int high)
{
    if (low < high)
    {
        // find the partition index
        int pi = partition(arr, low, high);

        // recursively sort elements before and after partition
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

int main()
{
    vector<int> arr = {10, 7, 8, 9, 1, 5};
    int n = arr.size();

    cout << "Original array: ";
    for (int x : arr)
        cout << x << " ";
    cout << endl;

    quickSort(arr, 0, n - 1);

    cout << "Sorted array: ";
    for (int x : arr)
        cout << x << " ";
    cout << endl;

    return 0;
}
